File: site/static/ansi.js
// ANSI to HTML: the string a terminal would receive, decoded into colored
// spans the way a terminal emulator would show it. This is how the live pages
// show a color mode honestly — the encoder's own SGR bytes, drawn — since the
// packed raster always carries the resolved color. The named colors take the
// xterm defaults, the same table site/src/ansi.rs uses for the static plates.

/// xterm's default rendering of the sixteen ANSI colors.
export const NAMED = [
  "#000000",
  "#cd0000",
  "#00cd00",
  "#cdcd00",
  "#0000ee",
  "#cd00cd",
  "#00cdcd",
  "#e5e5e5",
  "#7f7f7f",
  "#ff0000",
  "#00ff00",
  "#ffff00",
  "#5c5cff",
  "#ff00ff",
  "#00ffff",
  "#ffffff",
];

/// The xterm 256-color palette: the sixteen named colors, the 6×6×6 cube,
/// and the 24-step grey ramp.
export function ansi256(index) {
  if (index < 16) {
    return NAMED[index];
  }
  if (index >= 232) {
    const level = 8 + (index - 232) * 10;
    return `rgb(${level}, ${level}, ${level})`;
  }
  const cube = index - 16;
  const ramp = [0, 95, 135, 175, 215, 255];
  const r = ramp[Math.floor(cube / 36) % 6];
  const g = ramp[Math.floor(cube / 6) % 6];
  const b = ramp[cube % 6];
  return `rgb(${r}, ${g}, ${b})`;
}

const ESCAPES = {
  "&": "&amp;",
  "<": "&lt;",
  ">": "&gt;",
  '"': "&quot;",
  "'": "&#39;",
};

export function escapeHtml(text) {
  return String(text).replace(/[&<>"']/g, (character) => ESCAPES[character]);
}

/// A render clock for the readouts: two decimals under 10 ms, one under 100.
export function formatMs(ms) {
  if (ms < 10) {
    return `${ms.toFixed(2)} ms`;
  }
  if (ms < 100) {
    return `${ms.toFixed(1)} ms`;
  }
  return `${Math.round(ms)} ms`;
}

// The SGR state a run of text is drawn with.
function freshStyle() {
  return { foreground: null, background: null, bold: false };
}

function styleCss(style) {
  const parts = [];
  if (style.foreground) {
    parts.push(`color:${style.foreground}`);
  }
  if (style.background) {
    parts.push(`background:${style.background}`);
  }
  if (style.bold) {
    parts.push("font-weight:600");
  }
  return parts.join(";");
}

// Applies one SGR parameter list to `style`, in place. Handles reset, bold,
// the sixteen named colors on both channels, default-color resets, and the
// 256-color and truecolor extended forms.
function applySgr(style, codes) {
  if (codes.length === 0) {
    Object.assign(style, freshStyle());
    return;
  }
  for (let i = 0; i < codes.length; i++) {
    const code = codes[i];
    if (code === 0) {
      Object.assign(style, freshStyle());
    } else if (code === 1) {
      style.bold = true;
    } else if (code === 22) {
      style.bold = false;
    } else if (code >= 30 && code <= 37) {
      style.foreground = NAMED[code - 30];
    } else if (code >= 90 && code <= 97) {
      style.foreground = NAMED[code - 90 + 8];
    } else if (code >= 40 && code <= 47) {
      style.background = NAMED[code - 40];
    } else if (code >= 100 && code <= 107) {
      style.background = NAMED[code - 100 + 8];
    } else if (code === 39) {
      style.foreground = null;
    } else if (code === 49) {
      style.background = null;
    } else if (code === 38 || code === 48) {
      const channel = code === 38 ? "foreground" : "background";
      if (codes[i + 1] === 5 && codes.length > i + 2) {
        style[channel] = ansi256(codes[i + 2] & 0xff);
        i += 2;
      } else if (codes[i + 1] === 2 && codes.length > i + 4) {
        const r = codes[i + 2] & 0xff;
        const g = codes[i + 3] & 0xff;
        const b = codes[i + 4] & 0xff;
        style[channel] = `rgb(${r}, ${g}, ${b})`;
        i += 4;
      }
    }
  }
}

/// Decodes SGR-colored text into HTML: runs of text in spans carrying their
/// color as inline style, everything escaped. Other CSI sequences and OSC
/// strings (hyperlinks, images) are dropped.
export function ansiToHtml(text) {
  let html = "";
  let run = "";
  const style = freshStyle();
  const flush = () => {
    if (!run) {
      return;
    }
    const css = styleCss(style);
    html += css ? `<span style="${css}">${escapeHtml(run)}</span>` : escapeHtml(run);
    run = "";
  };
  let i = 0;
  while (i < text.length) {
    const character = text[i];
    if (character !== "\x1b") {
      run += character;
      i += 1;
      continue;
    }
    const next = text[i + 1];
    if (next === "[") {
      // CSI: parameters up to a final byte in 0x40–0x7e.
      let end = i + 2;
      while (end < text.length && !(text.charCodeAt(end) >= 0x40 && text.charCodeAt(end) <= 0x7e)) {
        end += 1;
      }
      if (text[end] === "m") {
        flush();
        const body = text.slice(i + 2, end);
        const codes = body === "" ? [] : body.split(";").map((code) => Number.parseInt(code, 10)).filter(Number.isFinite);
        applySgr(style, codes);
      }
      i = end + 1;
    } else if (next === "]") {
      // OSC: runs to BEL or ST (ESC \).
      let end = i + 2;
      while (end < text.length && text[end] !== "\x07" && !(text[end] === "\x1b" && text[end + 1] === "\\")) {
        end += 1;
      }
      i = text[end] === "\x07" ? end + 1 : end + 2;
    } else {
      i += 1;
    }
  }
  flush();
  return html;
}

File: site/src/ansi.rs
//! ANSI to HTML: the `String` a terminal would receive, decoded into colored
//! spans the way a terminal emulator would show it. The named colors take the
//! xterm defaults. This is how the site shows a color mode honestly — the
//! encoder's own SGR bytes, drawn — since the cards always carry RGB.

use crate::highlight::escape;

/// xterm's default rendering of the sixteen ANSI colors.
const NAMED: [&str; 16] = [
    "#000000", "#cd0000", "#00cd00", "#cdcd00", "#0000ee", "#cd00cd", "#00cdcd", "#e5e5e5",
    "#7f7f7f", "#ff0000", "#00ff00", "#ffff00", "#5c5cff", "#ff00ff", "#00ffff", "#ffffff",
];

fn ansi256(index: u8) -> String {
    if index < 16 {
        return NAMED[index as usize].to_string();
    }
    if index >= 232 {
        let level = 8 + (u32::from(index) - 232) * 10;
        return format!("rgb({level},{level},{level})");
    }
    let cube = u32::from(index) - 16;
    let ramp = [0, 95, 135, 175, 215, 255];
    let r = ramp[(cube / 36 % 6) as usize];
    let g = ramp[(cube / 6 % 6) as usize];
    let b = ramp[(cube % 6) as usize];
    format!("rgb({r},{g},{b})")
}

#[derive(Default, Clone, PartialEq, Eq)]
struct Style {
    foreground: Option<String>,
    background: Option<String>,
    bold: bool,
}

impl Style {
    fn css(&self) -> String {
        let mut parts = Vec::new();
        if let Some(color) = &self.foreground {
            parts.push(format!("color:{color}"));
        }
        if let Some(color) = &self.background {
            parts.push(format!("background:{color}"));
        }
        if self.bold {
            parts.push("font-weight:600".to_string());
        }
        parts.join(";")
    }

    fn apply(&mut self, codes: &[u16]) {
        let mut i = 0;
        while i < codes.len() {
            match codes[i] {
                0 => *self = Style::default(),
                1 => self.bold = true,
                22 => self.bold = false,
                30..=37 => self.foreground = Some(NAMED[(codes[i] - 30) as usize].to_string()),
                90..=97 => self.foreground = Some(NAMED[(codes[i] - 90 + 8) as usize].to_string()),
                40..=47 => self.background = Some(NAMED[(codes[i] - 40) as usize].to_string()),
                100..=107 => {
                    self.background = Some(NAMED[(codes[i] - 100 + 8) as usize].to_string())
                }
                39 => self.foreground = None,
                49 => self.background = None,
                38 | 48 => {
                    let target = if codes[i] == 38 {
                        &mut self.foreground
                    } else {
                        &mut self.background
                    };
                    match codes.get(i + 1) {
                        Some(5) => {
                            *target = codes.get(i + 2).map(|&n| ansi256(n as u8));
                            i += 2;
                        }
                        Some(2) => {
                            if let (Some(&r), Some(&g), Some(&b)) =
                                (codes.get(i + 2), codes.get(i + 3), codes.get(i + 4))
                            {
                                *target = Some(format!("rgb({r},{g},{b})"));
                            }
                            i += 4;
                        }
                        _ => {}
                    }
                }
                _ => {}
            }
            i += 1;
        }
    }
}

/// Decodes SGR-colored text into HTML spans; every other escape is dropped.
pub fn to_html(ansi: &str) -> String {
    let mut out = String::with_capacity(ansi.len());
    let mut style = Style::default();
    let mut open = false;
    let chars: Vec<char> = ansi.chars().collect();
    let mut i = 0;
    let mut text = String::new();
    let flush = |out: &mut String, text: &mut String, style: &Style, open: &mut bool| {
        if text.is_empty() {
            return;
        }
        let css = style.css();
        if css.is_empty() {
            out.push_str(&escape(text));
        } else {
            out.push_str("<span style=\"");
            out.push_str(&css);
            out.push_str("\">");
            out.push_str(&escape(text));
            out.push_str("</span>");
        }
        let _ = open;
        text.clear();
    };
    while i < chars.len() {
        if chars[i] == '\u{1b}' && chars.get(i + 1) == Some(&'[') {
            let mut end = i + 2;
            while end < chars.len() && !chars[end].is_ascii_alphabetic() {
                end += 1;
            }
            let body: String = chars[i + 2..end.min(chars.len())].iter().collect();
            if chars.get(end) == Some(&'m') {
                flush(&mut out, &mut text, &style, &mut open);
                let codes: Vec<u16> = body
                    .split(';')
                    .filter_map(|code| code.parse().ok())
                    .collect();
                if codes.is_empty() {
                    style = Style::default();
                } else {
                    style.apply(&codes);
                }
            }
            i = end + 1;
            continue;
        }
        text.push(chars[i]);
        i += 1;
    }
    flush(&mut out, &mut text, &style, &mut open);
    out
}
